### PR TITLE
Initial hw2 for Emma Clark (emmac4)

### PR DESCRIPTION
**Pull request by emmac4**

### CODE REVIEW

**Comment by calkins7** [comment]
Hi Emma,

Overall I think your code looks very clean and is easy to follow. Here are a few suggestions to improve readability of your code:
`train_discreteaction_pendulum.py`:
- Overall your hyperparameters look similar to the paper, but you may want to include a comment about why you used a batch size of 64 vs. 32 in the paper. 
- Plots look good!

`Algorithm.m`:
- Rename the `train_plot_avg`, or describe what it does in a comment within the function. It took me a few minutes to figure out how much it was doing when just looking at the functions calls in `train_discreteaction_pendulum.py`. 
- Line 129: It may help to make the size of the replay a variable you can pass into the DQN object.
- Lines 128-131: I was assuming that not using replay memory meant that we should make the replay buffer size 1, not size the batch size like your code currently has. Not sure which way is correct (or if it changes performance in the ablation study) but you may want to adapt these lines to allow for a batch size of 1 (or just pass in a batch size of 1 when you turn replay off).
- Line 135: I think we should use the RMSprop optimizer to be consistent with the paper.
- Line 143: you could specify a smaller initial experience replay buffer size as a separate input to your DQN object instead of filling the whole replay buffer at the start to save computation time (if that is an issue for your code). Also, to decrease run time, you could fill the initial buffer in the `rollout` function with random actions instead of doing the argmax on the initial policy NNs (which is essentially a random action rollout but querying the NN that much takes a long time sometimes).
- Line 170: the paper clipped the gradients at 1 instead of 100. 
- Lines 180 - 183: The paper decayed epsilon from 0.9 to 0.1 linearly over a large amount of simulations, then held epsilon constant at 0.1. The method you are using decays epsilon by 0.99 every 5 time steps. Small detail again, but may change performance! You may also want to make the coefficient (0.99) an input to the DQN so you can change it depending on how many simulations you are running. 

Overall your code was super easy to follow and very clean! These are all super small nitpicks. 

Grace

**Comment by emmac4** [comment]
Hi Grace, thanks for your comments. To address some of the things you said 

- I didn't end up using exactly the same hyperparameters as in the paper, the homework prompt said we may have to play around with these to find the right ones, I chose what seemed to give me the best results- the same goes for the optimizer, the RMSProp optimizer just didn't give as good of results as Adam 
- you were totally right about some of the function names being a bit vague, I should have fully commented the code before I opened the pull request, so that was my bad! I added some more description and rearranged a bit so that hopefully makes all of the functions clearer 
- for the replay buffer size, from my understanding when there is no replay buffer the max size should be the size of your batch, unless I misunderstood the homework prompt 
- From my what I understand the deque function I'm using for the replay is empty until you start appending to it and then it will delete from the beginning once you reach the max length, so I'm fairly certain this is not an issue with computation 
- I took your advice about random actions in the start, by just setting my initial epsilon value to 1.0 so the entire first iteration is random actions and this seemed to have a good impact! 
- for the gradient comment, I had been having some issues with the network not learning, so I was trying all kinds of different hyperparameters and gradient clipping values to try to fix it -- the problem ended up being completely unrelated to this but that was the reason the version I uploaded had the gradient clipped at 100; I have since fixed this and have it at 1 
- I went with exponential decay for epsilon just as a default since that is what I'm used to seeing in my limited RL experience, but I did end up changing it to linear decay because I thought it was easier to get epsilon to decay to the desired minimum value at a specified iteration number that way

I hope that covers all your comments/suggestions!
